lib/internal/streams/operators.js

```diff
@@ -0,0 +1,275 @@
+    concurrency = MathFloor(options.concurrency);
+  }
+
+  if (!NumberIsInteger(concurrency)) {
+    throw new ERR_INVALID_ARG_TYPE(
+      'concurrency', ['Integer'], concurrency);
+  }
+
+  if (concurrency < 1) {
+    throw new ERR_INVALID_ARG_VALUE('concurrency', concurrency);
+  }
+
+  const ac = new AbortController();
+  const stream = this;
+  const queue = [];
+  const signal = ac.signal;
+  const signalOpt = { signal };
+
+  options?.signal?.addEventListener('abort', () => ac.abort(), {
+    [kWeakHandler]: signalOpt
+  });
+
+  let next;
+  let resume;
+  let done = false;
+
+  function onDone() {
+    done = true;
+  }
+
+  async function pump() {
+    try {
+      for await (let val of stream) {
+        if (done) {
+          return;
+        }
+
+        if (signal.aborted) {
+          throw new AbortError();
+        }
+
+        try {
+          val = fn(val, signalOpt);
+        } catch (err) {
+          val = PromiseReject(err);
+        }
+
+        if (val === kEmpty) {
+          continue;
+        }
+
+        if (typeof val?.catch === 'function') {
+          val.catch(onDone);
+        }
+
+        queue.push(val);
+        if (next) {
+          next();
+          next = null;
+        }
+
+        if (!done && queue.length >= concurrency) {
+          await new Promise((resolve) => {
+            resume = resolve;
+          });
+        }
+      }
+      queue.push(kEof);
+    } catch (err) {
+      const val = PromiseReject(err);
+      val.catch(onDone);
+      queue.push(val);
+    } finally {
+      done = true;
+      if (next) {
+        next();
+        next = null;
+      }
+    }
+  }
+
+  pump();
+
+  try {
+    while (true) {
+      while (queue.length > 0) {
+        let val = queue[0];
+
+        if (val === kEof) {
+          return;
+        }
+
+        val = await val;
+
+        if (signal.aborted) {
+          throw new AbortError();
+        }
+
+        if (val !== kEmpty) {
+          yield val;
+        }
+
+        queue.shift();
+        if (resume) {
+          resume();
+          resume = null;
+        }
+      }
+
+      await new Promise((resolve) => {
+        next = resolve;
+      });
+    }
+  } finally {
+    ac.abort();
+
+    done = true;
+    if (resume) {
+      resume();
+      resume = null;
+    }
+  }
+}
+
+function filter(fn, options) {
+  return map.call(this, async (val) => {
+    return await fn(val) ? val : kEmpty;
+  }, options);
+}
+
+function drop(n, options) {
+  return map.call(this, (val) => {
+    n -= 1;
+    return n >= 0 ? kEof : val;
+  }, options);
+}
+
+function take(n, options) {
+  return map.call(this, (val) => {
+    n -= 1;
+    return n >= 0 ? val : kEof;
+  }, options);
+}
+
+function asIndexedPairs(options) {
+  let n = 0;
+  return map.call(this, (val) => {
+    return [n++, val];
+  }, options);
+}
+
+async function * flatMap(fn, options) {
+  for await (const val of map.call(this, fn, options)) {
+    if (isIterable(val)) {
+      yield * val;
```

Comment:
Do we need to check abort signal between yields?

Reply:
Probably not since `.return` propagates, if we need a `flatMapLatest` at some point it'd be different.